### DISCUSSION TITLE
AzureML Quickstart - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#5373] [Docs]
# AzureML Quickstart - Ultralytics YOLOv8 Docs

Step-by-step Quickstart Guide to Running YOLOv8 Object Detection Models on AzureML for Fast Prototyping and Testing

https://docs.ultralytics.com/guides/azureml-quickstart/

**Comment by helcionicolau** [comment]
Great!!!

**Reply by pderrenger** [reply]
@helcionicolau thank you! If you have any questions or need further assistance, feel free to reach out. Happy coding with YOLOv8! 😊🚀

**Comment by AnMaLeNo** [comment]
thank you !

**Reply by pderrenger** [reply]
@AnMaLeNo you're welcome! If you have any more questions or need further assistance with YOLOv8, feel free to reach out. Happy coding! 😊🚀

**Comment by rukshar69** [comment]
Very helpful. Are there fine-tuning-on-google-colab or Kaggle tutorials available?

**Reply by lmsasu** [reply]
Hi all,
While training the model, I get an error: 

torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 24.00 MiB. GPU 0 has a total capacty of 10.91 GiB of which 6.94 MiB is free. Including non-PyTorch memory, this process has 10.83 GiB memory in use. Of the allocated memory 10.51 GiB is allocated by PyTorch, and 39.52 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

This happens under Ubuntu 22.04 LTS, Python 3.11.5, pytorch 2.1.0, ultralytics 8.0.203, batch size=1, GPU NVIDIA 1080 Ti.
What should I do?

Thanks,
Lucian

**Reply by pderrenger** [reply]
@lmsasu hello Lucian,

It seems you're encountering an out-of-memory (OOM) error during training, which is common when the dataset or model is too large for the available GPU memory. Here are some steps to mitigate this issue:

1. **Reduce Batch Size**: You've mentioned using a batch size of 1, which is already minimal. However, if you're using multi-scale training or large image sizes, consider reducing the image size if possible.

2. **Use a Simpler Model**: If you're using a large YOLOv8 model, try switching to a smaller variant which requires less memory.

3. **Free Up GPU Memory**: Ensure no other processes are consuming GPU memory. You can check this with the `nvidia-smi` command.

4. **Gradient Accumulation**: If reducing the batch size is not an option, consider using gradient accumulation to effectively train with a larger batch size without increasing memory usage.

5. **Automatic Mixed Precision (AMP)**: Enable AMP to reduce memory usage and potentially speed up training with the `amp=True` parameter in your training script.

6. **Clear Cache**: Sometimes, PyTorch doesn't release unused memory. You can clear the cache with `torch.cuda.empty_cache()`.

7. **Adjust `max_split_size_mb`**: As suggested by the error message, you can set the `PYTORCH_CUDA_ALLOC_CONF` environment variable to a smaller `max_split_size_mb` to reduce fragmentation.

8. **Upgrade Hardware**: If none of the above solutions work, you may need to upgrade to a GPU with more memory or use model parallelism across multiple GPUs.

If you continue to face issues, please provide more details about your training configuration, and we can look into it further. 🛠️

**Comment by MOXHAN** [comment]
Hi, I want to make a validation run with a custom Yolov5 model using AzureML. But I dont want to do it through the environment of a compute instance. I want to create a job doing this. For this purpose, I created an AzureML environment based on the yolov5 docker image. However, when I run the python val.py command in the job, I get the following error:

from models.common import DetectMultiBackend
ModuleNotFoundError: No module named 'models'

Can somebody tell me what the reason for this and how to resolve it?

**Reply by MOXHAN** [reply]
The Model and Dataset being used are already registered in my workspace, I only upload the custom.yaml when submiting the job

**Reply by pderrenger** [reply]
@MOXHAN it seems like there might be an issue with the Python environment setup in your AzureML job. The error message indicates that the Python interpreter is unable to locate the `models` module, which should be part of the YOLOv5 repository.

Here are a few steps you can take to troubleshoot and resolve the issue:

1. **Verify the Working Directory**: Ensure that the working directory is set correctly in your AzureML job so that it aligns with the root of the YOLOv5 repository where the `models` directory is located.

2. **Check the Environment**: Make sure that the AzureML environment you've created from the YOLOv5 Docker image is correctly configured and includes all necessary dependencies. It should mimic the environment where YOLOv5 runs successfully.

3. **Dependencies**: Double-check that all required packages are installed in your environment. The YOLOv5 repository has a `requirements.txt` file that lists all the necessary Python packages. Ensure these are installed in your AzureML environment.

4. **Python Path**: Sometimes, the Python path may not include the current directory. You can explicitly add the repository root to the Python path in your job script using:
   ```python
   import sys
   sys.path.append('/path/to/yolov5')
   ```

5. **Container Structure**: If you're using a custom Docker container, ensure that the YOLOv5 code is correctly copied into the container and that the container's file structure is as expected.

6. **AzureML Job Script**: In your job submission script, ensure that you're referencing the correct paths for your model and dataset. Also, make sure that the `custom.yaml` file is being uploaded and referenced correctly.

7. **Logs**: Review the full logs of your AzureML job to see if there are any additional clues as to why the module is not being found.

8. **Test Locally**: If possible, try to run the same command locally in a Docker container based on the YOLOv5 image to see if you encounter the same issue.

Remember that while you're working with YOLOv5, this repository is focused on YOLOv8. If you're interested in using YOLOv8, you can follow similar steps to ensure your environment is correctly set up for YOLOv8 models. For more detailed information on setting up and using YOLOv8, you can refer to our documentation. 😊📚
